src/main/java/com/scalar/db/server/DistributedStorageService.java

```diff
@@ -0,0 +1,294 @@
+    execute(
+        () -> {
+          closeScanner(request.getScannerId());
+          responseObserver.onNext(Empty.newBuilder().build());
+          responseObserver.onCompleted();
+          return null;
+        },
+        responseObserver);
+  }
+
+  private void closeScanner(String scannerId) {
+    try {
+      ScannerHolder scannerHolder = scanners.remove(scannerId);
+      if (scannerHolder != null) {
+        scannerHolder.get().close();
+      }
+    } catch (IOException e) {
+      LOGGER.warn("failed to close the scanner. scannerId: " + scannerId);
+    }
+  }
+
+  @Override
+  public void mutate(MutateRequest request, StreamObserver<Empty> responseObserver) {
+    execute(
+        () -> {
+          List<Mutation> mutations = new ArrayList<>(request.getMutationsCount());
+          for (com.scalar.db.rpc.Mutation mutation : request.getMutationsList()) {
+            mutations.add(ProtoUtil.toMutation(mutation));
+          }
+
+          if (mutations.size() == 1) {
+            Mutation mutation = mutations.get(0);
+            if (mutation instanceof Put) {
+              storage.put((Put) mutation);
+            } else {
+              storage.delete((Delete) mutation);
+            }
+          } else {
+            boolean hasPut = false;
+            boolean hasDelete = false;
+            boolean mixed = false;
+            for (Mutation mutation : mutations) {
+              if (mutation instanceof Put) {
+                if (hasDelete) {
+                  mixed = true;
+                  break;
+                }
+                hasPut = true;
+              } else {
+                if (hasPut) {
+                  mixed = true;
+                  break;
+                }
+                hasDelete = true;
+              }
+            }
+
+            if (mixed) {
+              storage.mutate(mutations);
+            } else {
+              if (hasPut) {
+                storage.put(mutations.stream().map(m -> (Put) m).collect(Collectors.toList()));
+              } else {
+                storage.delete(
+                    mutations.stream().map(m -> (Delete) m).collect(Collectors.toList()));
+              }
+            }
+          }
+          responseObserver.onNext(Empty.newBuilder().build());
+          responseObserver.onCompleted();
+          return null;
+        },
+        responseObserver);
+  }
+
+  private static void execute(Callable<Void> callable, StreamObserver<?> responseObserver) {
+    try {
+      callable.call();
+    } catch (IllegalArgumentException | IllegalStateException e) {
+      LOGGER.error("an invalid argument error happened during the execution", e);
+      responseObserver.onError(
+          Status.INVALID_ARGUMENT.withDescription(e.getMessage()).asRuntimeException());
+    } catch (NoMutationException e) {
+      responseObserver.onError(
+          Status.FAILED_PRECONDITION.withDescription(e.getMessage()).asRuntimeException());
```

Comment:
For storage, it might be good enough to use gRPC status but I think we will need scalardb-specific error status for transaction. So, I'm wondering if we should have scalardb-specific error statuses for storage for consistency.
Maybe, it's something that is worth discussing.